### ISSUE TITLE
Add method returns a list on adding one file

### ISSUE THREAD

**Issue by khssnv** [ipfs/py-ipfs-api#133]
Adding one file user expects dict-type return (unfolded from original list-type response by [stream decorer](https://github.com/ipfs/py-ipfs-api/blob/master/ipfsapi/http.py#L119)), but if ipfs provider proceed request with error (even it's not critical) he got a list like its shown below.
```python
>>> ipfsapi.connect('http://127.0.0.1', 5001).add('./myfile')
[{'Name': 'myfile', 'Hash': 'QmbFMke1KXqnYyBBWxB74N4c5SBnJMVAiMNRcGu6x1AwQH', 'Size': '6'},
{'Message': 'cannot store pin state: write /var/lib/ipfs/datastore/000014.log: no space left on device', 'Code': 0, 'Type': 'error'}]
```
If user adding a directory he also get a list longer than expected.
```python
>>> ipfsapi.connect('http://127.0.0.1', 5001).add('./mydir')
[{'Name': 'mydir/myfile2', 'Hash': 'QmbFMke1KXqnYyBBWxB74N4c5SBnJMVAiMNRcGu6x1AwQH', 'Size': '6'},
{'Name': 'mydir/myfile1', 'Hash': 'QmbFMke1KXqnYyBBWxB74N4c5SBnJMVAiMNRcGu6x1AwQH', 'Size': '6'},
{'Name': 'mydir', 'Hash': 'Qmaqk3vCjaXMJf6HqeV7aTRvHZpELsJNRWU7WgFJ4Z9M9X', 'Size': '114'},
{'Message': 'cannot store pin state: write /var/lib/ipfs/datastore/000014.log: no space left on device', 'Code': 0, 'Type': 'error'}]
```
Maybe there is a better place for error messages to let `add` return expected type / size?

**Comment by whereswaldon** [comment]
How do you wish that it worked instead? We do need to propagate that error to the caller somehow, but I'd be open to hearing about alternatives for a future API.

**Comment by alexander255** [comment]
The plan for `ipfsapi` 0.5+ atm is that it will always return a list, making the return value more predictable. I do think however that we shouldn't just return these types of errors but actually raise an exception in this case (containing the partial result as a property). What do you think @whereswaldon @khssnv?

**Comment by whereswaldon** [comment]
I think it's appropriate to raise them as exceptions.

**Comment by alexander255** [comment]
Thank you! I cannot formally assign this to you on GitHub atm, but definitely fell free to work on this.

**Comment by alexander255** [comment]
Fixed by the exception handling part in #166 and making the number of return values explicitly defined in #167
